# 数字类型

## 概述

python语言提供的数字类型有：整数、浮点数、复数

## 整数

整数类型理论上是$[-\infty,\infty]$，实际上受限于运行程序的计算机的内存大小。可以使用pow(x,y)函数（用于计算$x^y$）来测试整数类型的取值范围

In [1]:
pow(2,100)

1267650600228229401496703205376

In [2]:
pow(2,500)

3273390607896141870013189696827599152216642046043064789483291368096133796404674554883270092325904157150886684127560071009217256545885393053328527589376

In [3]:
pow(2,pow(2,15))

ValueError: Exceeds the limit (4300 digits) for integer string conversion; use sys.set_int_max_str_digits() to increase the limit

In [5]:
import sys

# 增加整数字符串转换的最大位数限制
sys.set_int_max_str_digits(10000)  # 你可以根据需要调整这个值
pow(2,pow(2,15));#输出过长，省略输出结果

## 浮点数

python要求所有的浮点数必须带有小数部分，小数部分可以是零，这个设计可以用于区分整数和浮点数，浮点数有两种表示方法

1. 十进制表示法
2. 科学计数法
    - 96e4,4.3e-3,9.6E5
    - < a > e < b >=$a*10^b$
    
尽管浮点数0.0和整数0的含义相同，在计算机内部的表示方法也不同。

sys.float_info详细列出了python解释器运行系统的各项参数

In [6]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

上面输出结果中的参数：
- max：浮点所能表示的最大值
- max_exp：基数为2时最大值的幂
- max_10_exp：科学计数法最大值的幂
- dig：能准确计算的浮点数最大个数
- mant_dig：科学计数法中系数的最大精度
- epsilon：计算机所能分辨的两个相邻浮点数的最小差值

In [7]:
sys.float_info.max

1.7976931348623157e+308

浮点数在超过15位的数字计算中产生的误差与计算机采用二进制有关，浮点数无法进行极高精度的运算。python支持无限制且准确的整数计算，因此对于浮点数的运算可以将小数点去掉，当做整数来计算。例如

In [11]:
a,b=3.1415926535,1.234567898
print(a*b)
print(3.1415926535*1.234567898)
print(31415926535*1234567898)

3.8785094386037375
3.8785094386037375
38785094386037373430


#### decimal库的使用

In [13]:
import decimal
a=decimal.Decimal('3.141592653')
b=decimal.Decimal('1.234567898')
decimal.getcontext().prec=20
a*b

Decimal('3.878509437986453394')

浮点数类型的取值范围在$[2^{-1023},2^{1023}]$，运算精度为$2.220\times 10^{-16}$，对于除高精度科学计算外的绝大部分运算来说都够用

## 复数
python语言中复数的虚数部分用后缀“J”或者“j”表示，例如$12.3+4j,-5.6+7j,1.23e-4+5.67e+89j$

In [15]:
(1.23e-4+5.67e+89j).real

0.000123

In [16]:
(1.23e-4+5.67e+89j).imag

5.67e+89

In [2]:
round(0.14+0.25,2)

0.39

In [3]:
round(3.1415926)

3

In [4]:
round(3.1415926,2)

3.14

In [6]:
a,b,c,d,e,f='Python'

In [7]:
a

'P'

In [8]:
c

't'

In [9]:
x=12.34

In [10]:
type(x)

float

In [11]:
print(type(x))

<class 'float'>


In [12]:
5%6

5

In [13]:
6%2

0

In [14]:
4%3

1

In [15]:
3+5%6*2//8

4

In [16]:
10//8

1

In [17]:
18//8

2

In [20]:
a=input()
a[::-1]

123


'321'

In [21]:
a=eval(input())

123


In [22]:
a

123

In [25]:
print("{:->26}".format(a))

-----------------------123


In [28]:
for a in ["torch","soap","bath"]:
    print(a)

torch
soap
bath


In [29]:
for i in range(1,6):
    if i/3==0:
        break
    else:
        print(i,end=",")

1,2,3,4,5,

In [31]:
import random
random.random()

0.2644364879376949